lib/membrane_h264_plugin/parser.ex

```diff
@@ -255,32 +332,142 @@ defmodule Membrane.H264.Parser do
-      <<0, 0, 1, _rest::binary>> -> parameter_set
-      <<0, 0, 0, 1, _rest::binary>> -> parameter_set
-      parameter_set -> @prefix_code <> parameter_set
+  @spec initial_parameters_to_list(binary() | [binary()]) :: [binary()]
+  defp initial_parameters_to_list(pss) do
+    case pss do
+      <<>> -> []
+      ps when is_binary(ps) -> [ps]
+      pss -> pss
+    end
+  end
+
+  @spec get_incoming_parameter_sets(raw_stream_structure(), boolean(), state()) ::
+          {[binary()], [binary()]}
+  defp get_incoming_parameter_sets(:annexb, first_received_stream_format?, state) do
+    if first_received_stream_format?,
+      do: {state.initial_spss, state.initial_ppss},
+      else: {[], []}
+  end
+
+  defp get_incoming_parameter_sets({_avc, dcr}, _first_received_stream_format?, state) do
+    {:ok, %{spss: dcr_spss, ppss: dcr_ppss}} = DecoderConfigurationRecord.parse(dcr)
+
+    new_spss =
+      Enum.filter(dcr_spss, fn
+        sps -> sps not in Enum.map(state.cached_spss, fn {_id, nalu} -> nalu.payload end)
+      end)
+
+    new_ppss =
+      Enum.filter(dcr_ppss, fn
+        sps -> sps not in Enum.map(state.cached_ppss, fn {_id, nalu} -> nalu.payload end)
+      end)
+
+    {new_spss, new_ppss}
+  end
+
+  @spec stream_structures_compatible?(
+          raw_stream_structure() | stream_structure(),
+          raw_stream_structure() | stream_structure()
+        ) :: boolean()
+  defp stream_structures_compatible?(:annexb, :annexb), do: true
+  defp stream_structures_compatible?({avc, _}, {avc, _}), do: true
+  defp stream_structures_compatible?(_stream_structure1, _stream_structure2), do: false
+
+  @spec process_stream_format_parameter_sets([binary()], [binary()], CallbackContext.t(), state()) ::
+          {[Action.t()], state()}
+  defp process_stream_format_parameter_sets(
+         spss,
+         ppss,
+         ctx,
+         %{output_stream_structure: {:avc1, _}} = state
+       ) do
+    nalu_parser = state.nalu_parser
+
+    {parsed_spss, nalu_parser} = parse_nalus(spss, nalu_parser)
+    {parsed_ppss, nalu_parser} = parse_nalus(ppss, nalu_parser)
+
+    state = %{state | nalu_parser: nalu_parser}
+
+    new_spss = Enum.filter(parsed_spss, &(&1 not in Map.values(state.cached_spss)))
+    new_ppss = Enum.filter(parsed_ppss, &(&1 not in Map.values(state.cached_ppss)))
+
+    process_new_parameter_sets(new_spss, new_ppss, ctx, state)
+  end
+
+  defp process_stream_format_parameter_sets(spss, ppss, _ctx, state) do
+    frame_prefix = generate_frame_prefix(spss ++ ppss, state.input_stream_structure)
+    {[], %{state | frame_prefix: frame_prefix}}
+  end
+
+  @spec parse_nalus([binary()], NALuParser.t()) :: {[NALu.t()], NALuParser.t()}
+  defp parse_nalus(nalus, nalu_parser) do
+    Enum.map_reduce(nalus, nalu_parser, fn nalu, nalu_parser ->
+      NALuParser.parse(nalu, nalu_parser, false)
+    end)
+  end
+
+  @spec get_nalus_payloads([NALu.t()], stream_structure()) :: [binary()]
+  defp get_nalus_payloads(nalus, :annexb) do
```

Comment:
I would make it common with `NALUParser.unprefix_nalu_payload()` function so that not to repeat ourselves